### ISSUE TITLE
qml.qnn.KerasLayer does not support backprop in non-tape mode

### ISSUE THREAD

**Issue by iamlucaswolf** [PennyLaneAI/pennylane#944]
#### Issue description

Hey, me again, just noticed the following:

```python3
dev = qml.device('default.qubit.tf', wires=2)

@qml.qnode(dev, interface='tf')
def circuit(inputs, weights):
    qml.RX(inputs, wires=0)
    qml.Rot(*weights)
    return qml.expval(qml.PauliZ(0))
```

Now, when you try:

```python3
>>> qlayer = qml.qnn.KerasLayer(circuit, weight_shapes, output_dim=1)
AttributeError: 'PassthruQNode' object has no attribute '_dtype'
```

Notably, this only happens when specifying `interface='tf'` in the `qnode` decorator. 

* *Expected behavior:* (What you expect to happen)
N/A

* *Actual behavior:* (What actually happens)
N/A

* *Reproduces how often:* (What percentage of the time does it reproduce?)
Always.

* *System information:* (post the output of `import pennylane as qml; qml.about()`)
Name: PennyLane
Version: 0.13.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/iamlucaswolf/.cache/pypoetry/virtualenvs/quantum-rl-eji9J1nt-py3.8/lib/python3.8/site-packages
Requires: semantic-version, appdirs, autograd, numpy, toml, scipy, networkx
Required-by: 
Platform info:           Linux-5.4.0-53-generic-x86_64-with-glibc2.27
Python version:          3.8.6
Numpy version:           1.18.5
Scipy version:           1.5.4
Installed devices:
- default.gaussian (PennyLane-0.13.0)
- default.mixed (PennyLane-0.13.0)
- default.qubit (PennyLane-0.13.0)
- default.qubit.autograd (PennyLane-0.13.0)
- default.qubit.tf (PennyLane-0.13.0)
- default.tensor (PennyLane-0.13.0)
- default.tensor.tf (PennyLane-0.13.0)

#### Source code and tracebacks

Full stacktrace:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-27-90501f743a15> in <module>
----> 1 qlayer = qml.qnn.KerasLayer(circuit, weight_shapes, output_dim=1)

~/.cache/pypoetry/virtualenvs/quantum-rl-eji9J1nt-py3.8/lib/python3.8/site-packages/pennylane/qnn/keras.py in __init__(self, qnode, weight_shapes, output_dim, weight_specs, **kwargs)
    215         else:
    216             self._signature_validation(qnode, weight_shapes)
--> 217             self.qnode = to_tf(qnode, dtype=tf.keras.backend.floatx())
    218 
    219         # Allows output_dim to be specified as an int, e.g., 5, or as a length-1 tuple, e.g., (5,)

~/.cache/pypoetry/virtualenvs/quantum-rl-eji9J1nt-py3.8/lib/python3.8/site-packages/pennylane/interfaces/tf.py in to_tf(qnode, dtype)
     79     qnode_interface = getattr(qnode, "interface", None)
     80 
---> 81     if qnode_interface == "tf" and dtype == qnode._dtype:  # pylint: disable=protected-access
     82         return qnode
     83 

AttributeError: 'PassthruQNode' object has no attribute '_dtype'

#### Additional information
N/A

**Comment by co9olguy** [comment]
Thanks @iamlucaswolf for uncovering this. Just superficially, looks like probably be a really simple bug to fix (possibly just some logical error in the decorator). We'll take a look, but if you uncover the source in the meantime, keep us updated :smiley:

**Comment by iamlucaswolf** [comment]
Hi @co9olguy ,

From my understanding, the problem is that `KerasLayer` will call `pennylane.interfaces.tf.to_tf` on the given qnode, which seems to assume that all qnodes with interface "tf" have a `_dtype` member. While this is sometimes true for `TFQNode` (if the method gets bound during construction), it isn't for `PassthruQNode`, which I think makes sense  since `PassthruQNode` isn't tied to a specific interface and therefore doesn't necessarily deal with TensorFlow `dtypes` at all.

I guess adding a simple clause to the if statement to check whether the qnode has a `dtype` property is sufficient here? Also, since the `dtype` argument in `to_tf` defaults to `None` it probably would make sense to test whether the argument has been supplied at all and only then perform the check?

**Comment by josh146** [comment]
Hey @iamlucaswolf! I believe TensorFlow backpropagation with the `KerasLayer` class is something we only recently supported in PennyLane v0.13 (released end of last week!), and only in [tape mode](https://pennylane.readthedocs.io/en/latest/code/qml_tape.html).

This is partly for historical reasons:

* `default.qubit.tf` and TensorFlow backprop mode didn't exist when we first created `KerasLayer`.

* In the process of porting `KerasLayer` to support tape mode (finished only recently), the ability for `KerasLayer` to work with TensorFlow backprop was an unintentional (but very nice!) win.

> While this is sometimes true for TFQNode (if the method gets bound during construction), it isn't for PassthruQNode, which I think makes sense since PassthruQNode isn't tied to a specific interface and therefore doesn't necessarily deal with TensorFlow dtypes at all.

Yep, you are exactly right - in backprop mode, a TFQNode is not created, instead a PassthruQNode is created. In tape mode, these issues are removed --- there is only a single QNode class that is always used regardless of the differentiation method.

> I guess adding a simple clause to the if statement to check whether the qnode has a dtype property is sufficient here? Also, since the dtype argument in to_tf defaults to None it probably would make sense to test whether the argument has been supplied at all and only then perform the check?

I might get @trbromley to comment here, to see if this is something that we can quickly do to enable Keras backprop in non-tape mode. I'm not sure if there was another issue with how non-tape mode is architectured that made this difficult.

**Comment by josh146** [comment]
Just as an addendum, here is your example modified to work in tape mode:

```python
import pennylane as qml
import tensorflow as tf
import numpy as np

tf.keras.backend.set_floatx('float64')

qml.enable_tape()

dev = qml.device('default.qubit.tf', wires=2)

@qml.qnode(dev, interface='tf', diff_method="backprop")
def circuit(inputs, weights):
    qml.RX(inputs, wires=0)
    qml.Rot(*weights, wires=0)
    return qml.expval(qml.PauliZ(0))

qlayer = qml.qnn.KerasLayer(circuit, weight_shapes={"weights": (3,)}, output_dim=1)

with tf.GradientTape() as tape:
    res = qlayer(tf.constant([[0.543]], dtype=tf.float64))

grad = tape.gradient(res, qlayer.trainable_weights)

print("Output:", res)
print("Gradient:", grad)
```

With output:

```
Output: tf.Tensor([0.74493499], shape=(1,), dtype=float64)
Gradient: [<tf.Tensor: shape=(3,), dtype=float64, numpy=array([-3.43923588e-01, -4.84713776e-01,  5.55111512e-17])>]
```

**Comment by josh146** [comment]
@trbromley, curious what your thoughts are here. Should we attempt to backport backprop+Keras to non-tape mode? (In which case we should leave this issue open?).

Or mark it as 'won't fix', with the aim of making tape mode default soon, and closing this issue?

**Comment by trbromley** [comment]
Thanks @josh146, my preference would be to not spend time on additional features in non-tape mode, since they will quickly (hopefully?) become obsolete.

Although, I can't remember why we didn't support `PassThruQNode` in non-tape mode. You're probably right that it was for historical reasons, and if @iamlucaswolf would like to have a go at adding a quick fix then I'd be happy to look through!

**Comment by iamlucaswolf** [comment]
Hey @trbromley, sorry for the silence on my part; coursework kept me busy. 

I'll have a look at this after the 25th!

**Comment by trbromley** [comment]
Since we have just merged-in a PR to make tape mode default (https://github.com/PennyLaneAI/pennylane/pull/1040) and a release is coming shortly, we can close this PR. Non-tape mode will now be non-default and gradually removed.
